core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -949,14 +958,43 @@ private[chisel3] object Builder extends LazyLogging {
         case _ => None
       }.toSeq
 
+      /** Stores an adjacency list representation of groups.  Connections indicating children. */
+      val groupAdjacencyList = mutable
+        .LinkedHashMap[group.Declaration, mutable.LinkedHashSet[group.Declaration]]()
+        .withDefault(_ => mutable.LinkedHashSet[group.Declaration]())
+
+      // Populate the adjacency list.
+      groups.foreach { group =>
+        groupAdjacencyList(group.parent) = groupAdjacencyList(group.parent) += group
+      }
+
+      /** For a `group.Declaration`, walk all its children and fold them into a
+        * `GroupDecl`.  This "folding" creates one `GroupDecl` for each child
+        * nested under each parent `GroupDecl`.
+        */
+      def foldGroupDecls(decl: group.Declaration): GroupDecl = {
+        val children = groupAdjacencyList(decl)
+        val convention = decl.convention match {
+          case group.Convention.Bind => GroupConvention.Bind
+          case _                     => ???
+        }
+        decl match {
+          case leaf if children.isEmpty => GroupDecl(UnlocatableSourceInfo, leaf.name, convention, Seq.empty)
+          case node                     => GroupDecl(UnlocatableSourceInfo, node.name, convention, children.map(foldGroupDecls).toSeq)
```

Comment:
Should we not go ahead and capture the source locator in the `Declaration`?

Reply:
Yep. I was being lazy. I'll pipe this through and make sure it works. I'll do the same for group defines.

Reply:
Updated in 6f173786621b74c0ca52b3d2f006a34fb919755a.